src/eh1/delay.rs

```diff
@@ -1,23 +1,286 @@
+
+/// Delay transaction
+#[derive(PartialEq, Eq, Clone, Debug)]
+pub struct Transaction {
+    /// Kind is the transaction kind (and data) expected
+    kind: TransactionKind,
+    real_delay: bool,
+}
+
+/// Nanoseconds per microsecond
+const NANOS_PER_MICRO: u64 = 1_000;
+/// Nanoseconds per millisecond
+const NANOS_PER_MILLI: u64 = 1_000_000;
+
+impl Transaction {
+    /// Create a new delay transaction
+    pub fn new(kind: TransactionKind) -> Transaction {
+        Transaction {
+            kind,
+            real_delay: false,
+        }
+    }
+
+    /// Crete a new delay_ns transaction
+    pub fn delay_ns(ns: u32) -> Transaction {
+        Transaction::new(TransactionKind::Delay(ns.into()))
+    }
+
+    /// Crete a new delay_us transaction
+    pub fn delay_us(us: u32) -> Transaction {
+        Transaction::new(TransactionKind::Delay(us as u64 * NANOS_PER_MICRO))
+    }
+
+    /// Crete a new delay_ms transaction
+    pub fn delay_ms(ms: u32) -> Transaction {
+        Transaction::new(TransactionKind::Delay(ms as u64 * NANOS_PER_MILLI))
+    }
+
+    /// Create a new blocking delay_ns transaction
+    pub fn blocking_delay_ns(ns: u32) -> Transaction {
+        Transaction::new(TransactionKind::BlockingDelay(ns.into()))
+    }
+
+    /// Crete a new blocking delay_us transaction
+    pub fn blocking_delay_us(us: u32) -> Transaction {
+        Transaction::new(TransactionKind::BlockingDelay(us as u64 * NANOS_PER_MICRO))
+    }
+
+    /// Create new blocking delay_ms transaction
+    pub fn blocking_delay_ms(ms: u32) -> Transaction {
+        Transaction::new(TransactionKind::BlockingDelay(ms as u64 * NANOS_PER_MILLI))
+    }
+
+    /// Crete a new async delay_ns transaction
+    #[cfg(feature = "embedded-hal-async")]
+    pub fn async_delay_ns(ns: u32) -> Transaction {
+        Transaction::new(TransactionKind::AsyncDelay(ns.into()))
+    }
+
+    /// Crete a new async delay_us transaction
+    #[cfg(feature = "embedded-hal-async")]
+    pub fn async_delay_us(us: u32) -> Transaction {
+        Transaction::new(TransactionKind::AsyncDelay(us as u64 * NANOS_PER_MICRO))
+    }
+
+    /// Crete a new async delay_ms transaction
+    #[cfg(feature = "embedded-hal-async")]
+    pub fn async_delay_ms(ms: u32) -> Transaction {
+        Transaction::new(TransactionKind::AsyncDelay(ms as u64 * NANOS_PER_MILLI))
+    }
+
+    /// Perform an actual delay for this transaction
+    pub fn wait(mut self) -> Transaction {
+        self.real_delay = true;
+        self
+    }
+}
+
+/// MockDelay transaction kind.
+#[derive(PartialEq, Eq, Clone, Debug)]
+pub enum TransactionKind {
+    /// Any delay in nanoseconds, blocking or async. Should be a default kind for most use cases.
+    Delay(u64),
+    /// Delay in nanoseconds, must be a blocking delay. Expectation will fail for async delays.
+    BlockingDelay(u64),
+    /// Delay in nanoseconds, must be an async delay. Expectation will fail for blocking delays.
+    AsyncDelay(u64),
```

Comment:
Nit: The way rustdoc works, is that on overview pages (where multiple types are listed), the first line of the docstring is shown. Once you open the detail page, you see the full description.

This means that the following formatting might make it easier to navigate (plus some minor rewording):

```suggestion
    /// Expect any type of delay in nanoseconds
    ///
    /// The delay may be either blocking or async. In most cases, this is what you'll want to use.
    Delay(u64),
    /// Expect a blocking delay in nanoseconds
    ///
    /// The delay must be blocking. Expectation will fail for async delays.
    BlockingDelay(u64),
    /// Expect an async delay in nanoseconds
    ///
    /// The delay must be async. Expectation will fail for blocking delays.
    AsyncDelay(u64),
```

Reply:
Not really sure that I understand your comment about rustdocs. This is how both your and mine versions look to me:
![image](https://github.com/dbrgn/embedded-hal-mock/assets/73587400/ba706161-2797-4341-b07f-51b6d447711b)

![image](https://github.com/dbrgn/embedded-hal-mock/assets/73587400/951ec705-a9f9-492d-a6af-2e9362598c85)

![image](https://github.com/dbrgn/embedded-hal-mock/assets/73587400/8d4099f1-adab-4a37-adcc-4e69bbaf92d4)

![image](https://github.com/dbrgn/embedded-hal-mock/assets/73587400/8eb4de8d-4777-4bb1-b508-164ccfb039a1)

Both of them look identical to me. I would suggest that you do your preferred way of formatting and just push directly to my fork.

Reply:
It depends on the location where the text is displayed 🙂 As an example, the enum `ErrorKind` here is described using this format (first a short line with a summary, then a paragraph with more detailed explanations): https://docs.rs/embedded-hal/latest/embedded_hal/digital/enum.ErrorKind.html In the parent module, not the full text is displayed, but only the first paragraph: https://docs.rs/embedded-hal/latest/embedded_hal/digital/index.html As a consumer of docs, this helps a lot in my opinion to keep things tidy.

In the case of enum _variants_ I have to admit that right now there probably isn't an advantage in rustdoc. I did not realize that when writing my comment. However, I would still consider the pattern a best practice, just like with git commit messages: First line/paragraph is a short summary, optionally followed by further paragraphs that contain more details.

> Both of them look identical to me. I would suggest that you do your preferred way of formatting and just push directly to my fork.

If that's OK for you, I'll do that in this case!